src/Processors/QueryPlan/Optimizations/convertAnyJoinToSemiOrAntiJoin.cpp

```diff
@@ -0,0 +1,239 @@
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+
+#include <Columns/IColumn.h>
+#include <Common/logger_useful.h>
+#include <Common/Logger.h>
+// #include <Core/Block_fwd.h>
+#include <Core/ColumnsWithTypeAndName.h>
+#include <Core/Joins.h>
+#include <Interpreters/ActionsDAG.h>
+#include <Processors/QueryPlan/ExpressionStep.h>
+#include <Processors/QueryPlan/FilterStep.h>
+#include <Processors/QueryPlan/JoinStepLogical.h>
+
+namespace DB::QueryPlanOptimizations
+{
+
+namespace
+{
+
+ActionsDAG::IntermediateExecutionResult makeActionsInput(const ActionsDAG & dag, const ColumnsWithTypeAndName & input_columns)
+{
+    ActionsDAG::IntermediateExecutionResult input;
+
+    std::unordered_map<std::string_view, ColumnWithTypeAndName> output_map;
```

Comment:
The result of `evaluatePartialResult` is `ColumnsWithTypeAndName`, where some columns could be filled. Its size is equal to the number of outputs. It's done this way because duplicate names are allowed in ActionsDAG.

Here duplicate names will not work.